src/work/BatchWork.h

```diff
@@ -38,7 +38,7 @@ class BatchWork : public Work
 
     // Implementers aren't allowed to edit batching mechanism,
     // but instead provide iteration methods.
-    virtual bool hasNext() const = 0;
+    virtual bool hasNext() = 0;
```

Comment:
oh this is really ugly and error prone (`hasNext` with side effects)

Reply:
Yes, that’s why I didn’t want this work to inherit from BatchWork: there is no right place to check children works.